src/Illuminate/Validation/Concerns/ValidatesAttributes.php

```diff
@@ -1742,15 +1742,34 @@ public function requireParameterCount($count, $parameters, $rule)
     }
 
     /**
-     * Check if the parameters are of the same type.
+     * Check if the values to be compared are of the same type.
      *
-     * @param  mixed  $first
-     * @param  mixed  $second
+     * @param  string $attribute
+     * @param  mixed  $value
+     * @param  mixed  $comparedToValue
      * @return bool
      */
-    protected function isSameType($first, $second)
+    protected function isSameType($attribute, $value, $comparedToValue)
     {
-        return gettype($first) == gettype($second);
+        $requiredType = gettype($comparedToValue);
+
+        if (is_numeric($comparedToValue) && ! is_numeric($value)) {
+            $requiredType = 'numeric';
+        } elseif ($comparedToValue instanceof File && ! $value instanceof File) {
+            $requiredType = 'file';
+        }
+
+        if (gettype($value) != $requiredType) {
+            if (! in_array($requiredType, ['array', 'file', 'numeric', 'string'])) {
+                throw new InvalidArgumentException('The field under comparison must be an array, file, number or string.');
+            }
+
+            $this->addFailure($attribute, $requiredType);
```

Comment:
Why would we add a failure that is just the string "array" or "file", etc?

Reply:
The `addFailure` method accepts a `rule` argument. So the string we are passing is the `rule`. Since there are already rules and error messages set up for when the user input should be a `string`, `array`, `file` or `numeric`, I thought we could just use them to return proper error messages as opposed to creating new ones just for this use case. 

```
public function addFailure($attribute, $rule, $parameters = [])
    {
        if (! $this->messages) {
            $this->passes();
        }

        $this->messages->add($attribute, $this->makeReplacements(
            $this->getMessage($attribute, $rule), $attribute, $rule, $parameters
        ));

        $this->failedRules[$attribute][$rule] = $parameters;
    }
```